proto/job.proto

```diff
@@ -0,0 +1,101 @@
+// The file defines the CommandService API.
+syntax = "proto3";
+
+package job;
+
+option go_package = "/job";
+
+import "google/protobuf/timestamp.proto";
+
+// JobService provides methods for starting a job and streaming its output.
+service JobService {
+  // Starts the job and returns an identifier for the job.
+  rpc StartJob(StartRequest) returns (StartResponse) {}
+  // Returns a list of running jobs with their jobId and jobStart timestamps.
+  rpc ListJobs(ListJobsRequest) returns (ListJobsResponse) {}
+  // Stream the output of the job.
+  rpc StreamJobOutput(JobIdentifier) returns (stream JobOutput) {}
+  // Stops a running job.
+  rpc StopJob(JobIdentifier) returns (StopResponse) {}
+}
+
+// StartRequest contains the job command and rootFS to execute the command in.
+message StartRequest {
+  // The job command line to run.
+  string jobCommand = 1;
+  // The name of the root file system folder to use.
+  string rootFS = 2;
+  // A number representing the proportion of CPU shares the job has.
+  uint32 cpuShares = 3;
+  // Maximum memory limit in bytes.
+  uint64 memoryLimitInBytes = 4;
+  // A string in the form <device>:<io/s limit> (ex. /dev/sdc:1200) use to limit
+  // read I/O operations per second.
+  string diskReadIopsLimit = 5;
+  // A string in the form <device>:<io/s limit> (ex. /dev/sdc:1200) use to limit
+  // write I/O operations per second.
+  string diskWriteIopsLimit = 6;
+}
+
+// StartResponse is returned by StartJob if there is no error.
+message StartResponse {
+  // A unique string representing the running job.
+  string jobId = 1;
+  // The host system timestamp for when the job started.
+  google.protobuf.Timestamp jobStart = 2;
+}
+
+// JobIdentifier contains a jobId field. Set this field to the jobId value
+// returned when a job successfully starts.
+message JobIdentifier {
+  // A unique string representing the running job.
+  string jobId = 1;
+}
+
+// JobOutput contains the streamed output of the running job.
+message JobOutput {
+  string output = 1;
```

Comment:
How much data does `output` contain? Is `string` appropriate for arbitrary command output?

Reply:
To be honest, I'm not sure.

My instinct is to use string since I'm envisioning the output being mostly textual. If there was binary data involved, I'd lean toward bytes. Protobuf strings are more CPU intensive, I've read, due to UTF-8 validation, and they use more memory. They're human-readable, though, and a little easier to work with than bytes if there's encoding/decoding involved. Bytes would be easier if stream compression was desired. I understand neither can practically hold an infinite amount of data. I couldn't find a limit mentioned in the protobuf spec. Both might require some chunking if memory isn't assumed to be infinite.

I'll definitely try testing this, but I'm not certain and open to suggestions.